Bottom: 0ede322aa7cec8365db7b951fa66162ec1fafcdb
Top:    66cb3f35b7e4ca2e0294337a1e2345d056269d41
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-25 19:44:55 +0100

draw.c: Split "Gathering" from real drawing routines (Element names)


---

diff --git a/src/draw.c b/src/draw.c
index 7ad80f3..5af9fd1 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -189,6 +189,34 @@ Redraw (void)
 }
 
 static int
+element_callback (const BoxType * b, void *cl)
+{
+  ElementTypePtr element = (ElementTypePtr) b;
+  int *side = cl;
+
+  if (ON_SIDE (element, *side))
+    DrawElementPackage (element);
+  return 1;
+}
+
+static void
+draw_element_name (ElementType *element)
+{
+  if ((TEST_FLAG (HIDENAMESFLAG, PCB) && gui->gui) ||
+      TEST_FLAG (HIDENAMEFLAG, element))
+    return;
+  if (doing_pinout || doing_assy)
+    gui->set_color (Output.fgGC, PCB->ElementColor);
+  else if (TEST_FLAG (SELECTEDFLAG, &ELEMENT_TEXT (PCB, element)))
+    gui->set_color (Output.fgGC, PCB->ElementSelectedColor);
+  else if (FRONT (element))
+    gui->set_color (Output.fgGC, PCB->ElementColor);
+  else
+    gui->set_color (Output.fgGC, PCB->InvisibleObjectsColor);
+  DrawTextLowLevel (&ELEMENT_TEXT (PCB, element), PCB->minSlk);
+}
+
+static int
 name_callback (const BoxType * b, void *cl)
 {
   TextTypePtr text = (TextTypePtr) b;
@@ -199,7 +227,7 @@ name_callback (const BoxType * b, void *cl)
     return 0;
 
   if (ON_SIDE (element, *side))
-    DrawElementName (element);
+    draw_element_name (element);
   return 0;
 }
 
@@ -380,17 +408,6 @@ draw_element_package (ElementType *element)
   END_LOOP;
 }
 
-static int
-element_callback (const BoxType * b, void *cl)
-{
-  ElementTypePtr element = (ElementTypePtr) b;
-  int *side = cl;
-
-  if (ON_SIDE (element, *side))
-    draw_element_package (element);
-  return 1;
-}
-
 /* ---------------------------------------------------------------------------
  * prints assembly drawing.
  */
@@ -1451,19 +1468,11 @@ DrawElement (ElementTypePtr Element)
 void
 DrawElementName (ElementTypePtr Element)
 {
-  if (gui->gui && TEST_FLAG (HIDENAMESFLAG, PCB))
-    return;
+  assert (Gathering);
+
   if (TEST_FLAG (HIDENAMEFLAG, Element))
     return;
-  if (doing_pinout || doing_assy)
-    gui->set_color (Output.fgGC, PCB->ElementColor);
-  else if (TEST_FLAG (SELECTEDFLAG, &ELEMENT_TEXT (PCB, Element)))
-    gui->set_color (Output.fgGC, PCB->ElementSelectedColor);
-  else if (FRONT (Element))
-    gui->set_color (Output.fgGC, PCB->ElementColor);
-  else
-    gui->set_color (Output.fgGC, PCB->InvisibleObjectsColor);
-  DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), PCB->minSlk);
+  DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), 0);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1713,9 +1722,11 @@ EraseElementPinsAndPads (ElementTypePtr Element)
 void
 EraseElementName (ElementTypePtr Element)
 {
+  assert (Gathering);
+
   if (TEST_FLAG (HIDENAMEFLAG, Element))
     return;
-  DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), PCB->minSlk);
+  AddPart (&ELEMENT_TEXT (PCB, Element));
 }
 
 
@@ -1811,7 +1822,7 @@ static void
 draw_element (ElementTypePtr element)
 {
   draw_element_package (element);
-  DrawElementName (element);
+  draw_element_name (element);
   DrawElementPinsAndPads (element);
 }
